snuba/reader.py

```diff
@@ -31,6 +28,36 @@
 else:
     Result = MutableMapping[str, Any]
 
+
+def iterate_rows(result: Result) -> Iterator[Row]:
+    if "totals" in result:
+        return itertools.chain(result["data"], [result["totals"]])
+    else:
+        return iter(result["data"])
+
+
+def build_result_transformer(
+    column_transformations: Mapping[str, Callable[[Any], Any]],
+) -> Callable[[Result], None]:
+    """
+    Builds and returns a function that can be used to mutate a ``Result``
+    instance in-place by transforming all values for columns that have a
+    transformation function specified for their data type.
+    """
+
+    def transform_result(result: Result) -> None:
+        for column in result["meta"]:
+            transformer = column_transformations.get(column["type"])
+            if transformer is None:
+                continue
+
+            name = column["name"]
+            for row in iterate_rows(result):
+                row[name] = transformer(row[name])
```

Comment:
I think we could have a column name listed twice in result.meta (I remember something like that with aggregated time columns). In this case we would try to transform the same column twice thus throwing, that was prevented by line 70 in the previous implementation
https://github.com/getsentry/snuba/commit/30555b87e4fbe4570a77e6271cbbf36ca6bd6dd8

Reply:
Damn, good catch.

The test from that change stopped actually testing anything because the dates are static and the time range currently returns no rows. Cool. I was actually surprised there weren't any tests for this behavior to begin with, but coverage seemed good. I guess I was just looking in the wrong place.

Going to rewrite it to be a unit test against `transform_result` instead so that it's not subject to state or other query processing changes.

Reply:
Wait… reading through the result transformation code:
https://github.com/getsentry/snuba/blob/96fd71e081b2a017b8cb4d3e50f81d208492bb20/snuba/clickhouse/native.py#L141-L142

A column can be referenced in the `meta` sequence of the `Result` many times by the same name/alias (potentially with even varying types) as it is a sequence, but that same name  will only going to be included once per row in the `data` sequence of the result as it is a mapping (always the last value with that name as of Python 3.7, when we were running 2.7 this was actually nondeterministic.)

I don't think this is a problem with column transformation, but a design issue with the structure of the `Result` object (and thus the HTTP response body) itself. It's extremely confusing that the number of items in the `meta` sequence doesn't match the number of items (columns) in each row returned. The double formatting issue from #596 is just a symptom, not the problem.

Going to address that separately as a prerequisite for this change.

Reply:
> potentially with even varying types

I think that luckily I was wrong about this. The column name/alias seems to is unique (so the types and values will always be the same), but may be referenced multiple times in the result set. 
 The data returned will not be inconsistent, but we will be spending network on transmitting and CPU and memory on deserializing redundant data that will ultimately be discarded.

Reply:
Fixed (kinda) with #804.